### PR TITLE
Implementation of FallDistanceData

### PR DESCRIPTION
**Pull request by ZephireNZ**
SpongePowered/SpongeAPI#891

Implementation of `FallDistanceData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `ImmutableFallDistanceData`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `FallDistanceValueProcessor`(s):
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `FALL_DISTANCE` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code:

```java
@Plugin(id = "data-test", name = "DataTest")
public class DataTest {

    @Listener
    public void onDamage(DisplaceEntityEvent.Move.TargetLiving event) {
        Entity entity = event.getTargetEntity();
        if(entity.supports(FallDistanceData.class) && entity instanceof Player) {
            FallDistanceData data = entity.getOrCreate(FallDistanceData.class).get();
            data.set(Keys.FALL_DISTANCE, 0F);
            entity.offer(data);
        }
        if(entity instanceof Player) {
            float dist = entity.get(Keys.FALL_DISTANCE).get();
            System.out.println(((Player) entity).getProfile().getName() + " fell " + dist);
        } else {
            Optional<MutableBoundedValue<Float>> opt = entity.getValue(Keys.FALL_DISTANCE);
            if(opt.isPresent()) {
                System.out.println(entity.getUniqueId().toString() + " fell " + opt.get().get());
            }
        }
    }
}
```

Notes:
Your test plugin should test using all possible value related methods with your `DataManipulator`s and also test the serialization and deserialization of your `DataManipulator`s.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/processor/data/entity/FallDistanceDataProcessor.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+public class FallDistanceDataProcessor extends AbstractEntitySingleDataProcessor<EntityLivingBase, Float, MutableBoundedValue<Float>,
+        FallDistanceData, ImmutableFallDistanceData> {
+
+    public FallDistanceDataProcessor() {
+        super(EntityLivingBase.class, Keys.FALL_DISTANCE);
+    }
+
+    @Override
+    protected boolean set(EntityLivingBase entity, Float value) {
+        entity.fallDistance = value;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/FallDistanceDataProcessor.java]: Needs some validation on the value itself.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/FallDistanceValueProcessor.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                .actualValue(value)
+                .defaultValue(0F)
+                .minimum(0F)
+                .maximum(Float.MAX_VALUE)
+                .build();
+    }
+
+    @Override
+    protected boolean set(EntityLivingBase container, Float value) {
+        container.fallDistance = value;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/FallDistanceValueProcessor.java]: And here. needs some validation.

**Comment by ZephireNZ** [comment]
The test code is annoying - tried it with damage event, nope doesn't seem implemented. Same goes with move even, but *only* the player one is implemented -_-

**Comment by ZephireNZ** [comment]
Oopsy xD

**Comment by ZephireNZ** [comment]
Looks like there's some duplication between this and #234 - refactoring is gonna be soo fun...

**Comment by gabizou** [comment]
@ZephireNZ Yeah.... considering that PR attempts to solve a lot of problems at the same time, it might take some time at the least. It might be divided into phases as @Mumfrey and I have pondered on some ways to resolve the issue of "stateless" `CatalogType` references while at the same time revamping the overall `SpongeGameRegistry` as a whole.

**Comment by ZephireNZ** [comment]
Wait derp, wrong PR to put that with xD

**Comment by gabizou** [comment]
@ZephireNZ to save time, I'm gonna re-post the above comment in #217

**Pull request closed without merging**